lib/ads.js

```diff
@@ -146,14 +161,24 @@ var end = function (cb) {
   releaseSymHandles.call(ads, function () {
     releaseNotificationHandles.call(ads, function () {
       if (ads.tcpClient) {
-        // ads.tcpClient.end(); // reijo removed
-        ads.tcpClient.destroy() // reijo added
+        ads.tcpClient.destroy()
       }
       if (cb !== undefined) cb.call(ads)
     })
   })
 }
 
+var endNoConn = function (cb) {
+  var ads = this
+  ads.tcpClient.removeListener('data', ads.dataCallback)
+  releaseNotificationHandles.call(ads, function () {
```

Comment:
releaseSymHandles out
releaseNotificationHandles in?

Reply:
Compare it with end -> only releaseSymHandles are out ... The reason vor that is relaseSymHandles need communication to PLC

Reply:
releaseNotificationHandles calls deleteDeviceNotificationCommand
```
var deleteDeviceNotificationCommand = function (notificationHandle, cb) {
  var buf = Buffer.alloc(4)
  buf.writeUInt32LE(notificationHandle, 0)

  var options = {
    commandId: ID_DEL_NOTIFICATION,
    data: buf,
    cb: cb
  }
  runCommand.call(this, options)
```
and it runs a command on the PLC :-)

Reply:
Another question is the Handles in the following arrays will not released. 
Must this on the next connection or can we delete the array items.
  ads.symHandlesToRelease = []
  ads.notificationsToRelease = []

Reply:
F... yes u are right ... Im to fast and not really thinking about it ... Im wondering now wy it works for me when i delete only the releaseSymHandles? ... Anyway so i think we should delete it correctly here -> 

```
var endNoConn = function (cb) {
  var ads = this

  ads.tcpClient.removeListener('data', ads.dataCallback)
  ads.symHandlesToRelease = []
  ads.notificationsToRelease = []

  if (ads.tcpClient) {
    ads.tcpClient.destroy()
  }

  if (cb !== undefined) cb.call(ads)
}
```

What do you think about that?

Reply:
Why does it have to be controlled from the outside? I have a member for the Node-Red implementation which always has problems when switching off the PLC. At your suggestion, I have inserted a property connected in the node-ads-api and ask them when releasing the trade from. This leaves the decision within the API. If you want, you can make a diff to your implementation.

Reply:
Yes i see ... Thats a nice solution ... But can u help me how this can work? -> The cb would not be called when the if of ads.connected not true ... Same on deleteDeviceNotificationCommand 

```
var releaseSymHandles = function (cb) {
  var ads = this
  if (this.symHandlesToRelease.length > 0) {
    var symHandle = this.symHandlesToRelease.shift()
    releaseSymHandle.call(this, symHandle, function () {
      releaseSymHandles.call(ads, cb)
    })
  } else cb.call(this)
}


var releaseSymHandle = function (symhandle, cb) {
  var ads = this
  if (ads.connected) {
    var commandOptions = {
      indexGroup: ADSIGRP.RELEASE_SYMHANDLE,
      indexOffset: 0x00000000,
      bytelength: symhandle.length,
      bytes: symhandle
    }
    writeCommand.call(this, commandOptions, function (err) {
      cb.call(ads, err)
    })
  }
}
```

Reply:
Oh F**k.. 
You are right, no problem 
calling the cb on else will correct this.
#*!?!!*#

Reply:
Okay i have merged your changes to my Fork but not tested at this Time - i can do it today in the evening... So it´s nearly the same with your Fork - I killed my changes endNoConn it should be not necessary anymore

I will give feedback when i finish with the tests ... Thanks for your Support

Reply:
This Version works as expected :-)